### PR TITLE
JPA configuration changes

### LINKED ISSUE #2568: Review schema requirements for classic mode
We will retrieve this information from Skipper repository instead.

child of #2548

### LINKED ISSUE #2596: Basic jpa structure for domain classes
- Convert JPA classed to Domain entities
- Remove explicit RDBMS based repository implementation
- Relax tightly coupled validation rules

child of #914

### PR DESCRIPTION
**Pull request by ilayaperumalg**
- Move Stream/Task Definition repositories as JPA repositories
  - Update repository method names in align with the entity field names
  - Update existing Stream/Task definition domain classes into JPA persistent entity classes
    - Update the corresponding database column names and transient fields
  - Update DataFlowControllerConfiguration to have the corresponding EntityScan/EnableJPARepositories/EnableMapRepositories
  - Remove existing conditional check on StreamDefinitionRepository when stream feature is enabled
  - Remove existing RDBMS/Inmemory repository implementations for stream/task definition repositories
  - Remove unused DeploymentIdRepository and its implementations
  - Move Audit repository code into server-core repository package
  - Move Audit domain classes from `spring-cloud-dataflow-server-core` into `spring-cloud-dataflow-core`
  - Fix Pageable search query using repository method
  - Update test configuration classes
     - Remove explicit DataSourceTransactionManager and use JPATransactionManager for tests
       - This is needed for making the active transaction session inside JPASession

Resolves #2596
Resolves #2568

### CODE REVIEW

**Comment by jvalkeal** [comment]
I'm getting this:
```
[ERROR] Errors: 
[ERROR]   DefaultStreamServiceUpdateTests.testCreateUpdateRequestsWithoutRegisteredApp:89->testCreateUpdateRequests:103 » InvalidDataAccessResourceUsage
[INFO] 
[ERROR] Tests run: 308, Failures: 0, Errors: 1, Skipped: 2
```
PR were behind one commit from master so not sure if it's related to that.

File: spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamDefinition.java [hunk 1]
```diff
@@ -112,7 +136,14 @@ public String getDslText() {
 	 * @return iterator that iterates over the application definitions in deployment order
 	 */
 	public Iterator<StreamAppDefinition> getDeploymentOrderIterator() {
-		return new ReadOnlyIterator<>(this.applicationDefinitions.descendingIterator());
+		return new ReadOnlyIterator<>(getAppDefinitions().descendingIterator());
+	}
+
+	@PostLoad
+	public void initialize() {
+		if (this.applicationDefinitions == null || this.applicationDefinitions.isEmpty()) {
```
Comment (tzolov) [inline spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/StreamDefinition.java]: Consider the `CollectionUtils.isEmpty(this.applicationDefinitions)` instead

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java [hunk 2]
```diff
@@ -185,7 +192,7 @@ public AuditRecordController auditController(
 	}
 
 	@Bean
-	@ConditionalOnBean({ StreamDefinitionRepository.class })
+	@ConditionalOnStreamsEnabled
```
Comment (tzolov) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: Nice improvement. 
Perhaps we should group all ConditionalOnStreamsEnabled beans in a common inner class and annotate only that class.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java [hunk 3]
```diff
@@ -206,20 +213,20 @@ public MetricStore metricStore(MetricsProperties metricsProperties) {
...
 	@Bean
-	@ConditionalOnBean(StreamDefinitionRepository.class)
+	@ConditionalOnStreamsEnabled
 	public StreamValidationController streamValidationController(StreamService streamService) {
 		return new StreamValidationController(streamService);
 	}
 
 	@Bean
-	@ConditionalOnBean(TaskDefinitionRepository.class)
+	@ConditionalOnTasksEnabled
```
Comment (tzolov) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: Consider grouping all ConditionalOnTasksEnabled beans in a common inner configuration class.

File: spring-cloud-dataflow-core/pom.xml [hunk 4]
```diff
@@ -25,6 +25,26 @@
 			<groupId>commons-lang</groupId>
 			<artifactId>commons-lang</artifactId>
 		</dependency>
+		<dependency>
+			<groupId>org.hibernate.javax.persistence</groupId>
+			<artifactId>hibernate-jpa-2.1-api</artifactId>
+		</dependency>
+        <dependency>
+            <groupId>org.springframework.data</groupId>
+            <artifactId>spring-data-jpa</artifactId>
```
Comment (jvalkeal) [inline spring-cloud-dataflow-core/pom.xml]: spaces vs. tabs

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java [hunk 5]
```diff
@@ -211,16 +210,16 @@ public SimpleJobServiceFactoryBean simpleJobServiceFactoryBean(DataSource dataSo
...
 		repositoryFactoryBean.setDataSource(dataSource);
-		repositoryFactoryBean.setTransactionManager(dataSourceTransactionManager);
+		repositoryFactoryBean.setTransactionManager(platformTransactionManager);
 		return repositoryFactoryBean;
 	}
 
 	@Bean
-	public DataSourceTransactionManager transactionManagerForServer(DataSource dataSource) {
-		return new DataSourceTransactionManager(dataSource);
+	public PlatformTransactionManager transactionManager(EntityManagerFactory entityManagerFactory) {
```
Comment (jvalkeal) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java]: This is probably not needed anymore as it should come from boot. At latest it's going to get removed when we go to boot21. We added custom when we were on boot14 and boot15 added manager config.

Comment (ilayaperumalg) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java]: Yeah, probably once we update to Boot 2.1.x. Currently, it is needed to make the transaction active inside the tests.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java (hunk 2 above)
Comment (ilayaperumalg) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: @tzolov yeah, that would be nice! Let's do this in a different story. I will create one for it.

**Comment by jvalkeal** [comment]
merged per 23352835164eb710802a20e7738e8776ec15198d

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java [hunk 6]
```diff
@@ -149,11 +152,15 @@
 @EnableCircuitBreaker
 @EntityScan({
 		"org.springframework.cloud.dataflow.registry.domain",
-		"org.springframework.cloud.dataflow.server.audit.domain"
+		"org.springframework.cloud.dataflow.core"
+})
+@EnableMapRepositories(basePackages = {
```
Comment (markpollack) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/config/DataFlowControllerAutoConfiguration.java]: Why is this here - there are no map repositories?
